### PR TITLE
add support for empty microindexes

### LINKED ISSUE #987: zar index leaves empty files upon indexing error
This was described in https://github.com/brimsec/zq/issues/815#issuecomment-651423326. Splitting out into its own bug:

```sh
$ zar import multitype.tzng 

$ zar index orig_h
/Users/phil/logs/20091119/1258594908.85978.zng: creating index /Users/phil/logs/20091119/1258594908.85978.zng.zar/zdx-field-orig_h
type of orig_h field changed from string to ip

$ tree -s $ZAR_ROOT
/Users/phil/logs
├── [        128]  20091119
│   ├── [         93]  1258594908.85978.zng
│   └── [         96]  1258594908.85978.zng.zar
│       └── [          0]  zdx-field-orig_h.zng
└── [        197]  zar.json

2 directories, 3 files

$ zar find orig_h=192.168.2.1
/Users/phil/logs/20091119/1258594908.85978.zng.zar/zdx-field-orig_h: /Users/phil/logs/20091119/1258594908.85978.zng.zar/zdx-field-orig_h: cannnot read zdx header
```

### PR DESCRIPTION
**Pull request by mccanne**
This commit adds support for the notion of a valid microindex with
nothing in it.  This is useful for knowing when something has been
indexed but there are no matching elements.  An empty microindex
is a zng file with a microinidex trailer with various null fields.

We also fixed a bug where zar index would leave behind an empty file
(which is not a valid microindex) when it encountered an error.
The fix is to add an Abort() method on the microindex Writer that can
be called in place of Close() to cleanup any partial state.

Fixes #987

### CODE REVIEW

File: zdx/writer.go [hunk 1]
```diff
@@ -17,23 +17,34 @@ import (
...
 // as zng records are consumed.
 //
 // The keyFields argument to NewWriter provides a list of the key names, ordered by
 // precedence, that will serve as the keys into the index.  The input
 // records may or may not have all the key fields.  If a key field is
-// missing it appears as a null value in the index.  Nulls are sorted
-// before all non-null values.
+// missing, it appears as a null value in the index.  Nulls are sorted
+// before all non-null values.  All key fields must have the same type.
+// The Writer may detect an error if a key fiield changes type but does not
```
Comment (alfred-landrum) [inline zdx/writer.go:28]:
```suggestion
// The Writer may detect an error if a key field changes type but does not
```

File: zdx/writer.go [hunk 2]
```diff
@@ -17,23 +17,34 @@ import (
...
 // with the precedence order of the keyFields.
 //
-// As the zng file data is written, a b-tree index is computed as a
-// constant b-tree to make key lookups efficient.  The b-tree sections
+// As the zng file data is written, a B-tree index is computed as a
+// constant B-tree to make key lookups efficient.  The b-tree sections
 // are written to temporary files and at close, they are merged into
 // a single-file microindex.
+//
+// If a Writer is created the Closed without ever writing records to it, then
```
Comment (alfred-landrum) [inline zdx/writer.go:41]:
```suggestion
// If a Writer is created but Closed without ever writing records to it, then
```

File: zdx/writer.go [hunk 3]
```diff
@@ -17,23 +17,34 @@ import (
...
+// the index is created with no keys and an "empty" microindex trailer.  This is
+// useful for knowing when something has been indexed but no keys were present.
+// If a Writer is created then an error is enountered (for example, the type of)
```
Comment (alfred-landrum) [inline zdx/writer.go:44]: this sentence left me hanging

File: zdx/writer.go [hunk 4]
```diff
@@ -17,23 +17,34 @@ import (
...
+// this type consistency.  For example, the caller should create a separate
+// microindex for fields that have a common name but different types.
 //
 // The keys in the input zng stream must be previously sorted consistent
 // with the precedence order of the keyFields.
 //
-// As the zng file data is written, a b-tree index is computed as a
-// constant b-tree to make key lookups efficient.  The b-tree sections
+// As the zng file data is written, a B-tree index is computed as a
+// constant B-tree to make key lookups efficient.  The b-tree sections
```
Comment (nwt) [inline zdx/writer.go:37]:
```suggestion
// constant B-tree to make key lookups efficient.  The B-tree sections
```

File: zdx/writer.go [hunk 5]
```diff
@@ -108,22 +122,44 @@ func (w *Writer) Write(rec *zng.Record) error {
...
-		// encounted, just close the underlying storage object and return.
-		return w.iow.Close()
+	firstErr := w.closeTree()
+	if err := w.iow.Close(); err != nil && firstErr == nil {
+		firstErr = err
 	}
+	if err := iosrc.Remove(w.uri); err != nil && firstErr == nil {
+		firstErr = err
+	}
+	return firstErr
```
Comment (nwt) [inline zdx/writer.go:137]:
```suggestion
	firstErr := w.closeTree()
	if err := w.iow.Close(); firstErr == nil {
		firstErr = err
	}
	if err := iosrc.Remove(w.uri); firstErr == nil {
		firstErr = err
	}
	return firstErr
```

File: zdx/writer.go [hunk 6]
```diff
@@ -108,22 +122,44 @@ func (w *Writer) Write(rec *zng.Record) error {
...
+	}
+	if w.writer == nil {
+		// If the writer hasn't been created because no records were
+		// encountered, then the base layer writer was never created.
+		// In this case, bypass the base layer, write an empty trailer
+		// directly to the output, and close.
+		err := w.writeEmptyTrailer()
+		err2 := w.iow.Close()
+		if err == nil {
+			err = err2
```
Comment (nwt) [inline zdx/writer.go:158]:
```suggestion
		if err2 := w.iow.Close(); err == nil {
			err = err2
```

File: zdx/writer.go [hunk 7]
```diff
@@ -139,6 +175,19 @@ func (w *Writer) Close() error {
 	return w.writer.buffer.Close()
 }
 
+func (w *Writer) closeTree() error {
+	if w.writer == nil {
+		return nil
+	}
+	var firstErr error
+	for p := w.writer.parent; p != nil; p = p.parent {
+		if err := p.Close(); err != nil && firstErr == nil {
```
Comment (nwt) [inline zdx/writer.go:184]:
```suggestion
		if err := p.Close(); firstErr == nil {
```

**Pull request merged**